### PR TITLE
add event type as tag to measurements sent to DB to discriminate regular from soft polled events

### PR DESCRIPTION
**Pull request by iamtrep**

### CODE REVIEW

**Comment by iamtrep** [comment]
There is a potential side effect with adding a new column in a stream, depending on how the graph is configured it can make it look like two different measurements, change the line color etc.   So perhaps it would be best to make this a preference.

Otherwise finding it really useful to be able to discriminate device events from polled state snapshots.

Please advise.

**Comment by Mavrrick** [comment]
I was looking at this after one of the posts in the support thread. I took a slightly different approach. 

Instead of creating a totaly new set of data. It seemed very reasonable to use an existing column of the event data that is already present and enable it to capture valid data for real events and then lean on it to also handle the polled identifier value. 

I added:  
String type = escapeStringForInfluxDB(evt.type) on line 386 

and updated like 391 to     
String data = "${measurement},deviceId=${deviceId},deviceName=${deviceName},hubName=${hubName},locationName=${locationName},type=${type}"

That enabled the ability to start to capture the Type value passed with events. This often has the Digital or physical tag.

Then added
type: "softpoll", on line 628 and 651

This makes the softpoll label the Type value as softpoll allowing the data to be identified in that situation.

**Comment by iamtrep** [comment]
Yeah I came to the same conclusion as you... just pushed the commits I had locally.  Reason I changed the commit title is that, as adding this data as a field (instead of a tag) doesn't create a new series, it won't impact current users.

I don't like soft polling as an expression so I used "sampleType" and set it to either "state"  (which is what it is) or "event".  Up to you guys of course.

Really like the idea of adding the physical/digital event type as well !   I noticed jtp10181/Jeff Page adds this as a tag, not a field. 
 
> This makes the softpoll label the Type value as softpoll allowing the data to be identified in that situation.

Ah I see - so your idea is that the new field (which I called "sampleType" in my case) can take on values "digital", "physical" or "state"(/"softPolled") ?  Makes sense.

I gotta say I'm a little bit out of my depth with InfluxDB, I spent a couple of hours trying to understand the differences between field and tag (indexed or not, changes often/rarely, etc) but I am definitely not an InfluxDB power user.

File: influxdb-logger.groovy [hunk 1]
```diff
@@ -582,6 +582,9 @@ def handleEvent(evt, softPolled = false) {
         data += ",unit=${unit} value=${value}"
     }
 
+    // add sample type to the measurement as field
+    data += ",sampleType=\"${sampleType}\""
```
Comment (Mavrrick) [inline influxdb-logger.groovy]: I wanted to ask why did you remove your sampleType value from line 394 and add it here. It seems like it should be on 394 as it is part of the information passed with the Event value that is parsed in that block of code.

**Comment by iamtrep** [comment]
May need more work as some events have evt.type == null (e.g. power reporting).  If we're ok with type=null, then it's good to go.

File: influxdb-logger.groovy (hunk 1 above)
Comment (iamtrep) [reply influxdb-logger.groovy]: See comments near the top of handleEvent():
`// Build data string to send to InfluxDB:`
 `//  Format: <measurement>[,<tag_name>=<tag_value>] field=<field_value>`
 `//    If value is an integer, it must have a trailing "i"`
 `//    If value is a string, it must be enclosed in double quotes.`
In fact for completeness the format spec should read:

`//  Format: <measurement>[,<tag_name>=<tag_value>] field=<field_value> timestamp`

The difference is that having it up there with deviceName et al. makes it par of the list of tags.   I've put it in with the fields (cf. my hesitations in previous comment).

Comment (iamtrep) [reply influxdb-logger.groovy]: So I guess I am confused as to what you've actually done. If you want to point me to a commit I can look at it, would be easier.

**Comment by Mavrrick** [comment]
Event type is frequently left blank for allot of devices. So null wouldn't be unexpected.

**Comment by Mavrrick** [comment]
Can you please elaborate on what you mean by the impact of putting that value in the tag section instead of field section. I implemented my change and did not see any difference in my graphs. I know that may not be true for all folks,  but I didn't see anything.

When trying to research tag vs field with InfluxDB it seems to be largely related to indexing. Tags are indexed while fields are not. It also seems there are query limitations to fields that tags don't. 

So the reason using the type as a tag is:
1. You will index based on the limited number of values it will likely have. As of right now that is: null, Physical, Digital, SoftPoll
2. You will be able to query using the Tag in a where or group by clause of a query. 
3. Because it is indexed it will help significantly improve performance of retrieving data.

If this field was going to have hundreds of potential values then it may be better as a field instead of tag, but since it only has a few values this seems like the appropriate way to go.

**Comment by dennypage** [comment]
Setting aside the potential effect for things like Grafana, I would view the distinction of real vs artificial event as an attribute of the data (metadata) rather than data itself. Like the hub name and device id/name that produced the measurement. In other words, I view it as a tag rather than a value.

**Comment by iamtrep** [comment]
I tend to agree with you guys on the semantics. 

I’ll give you a couple of examples of the issues I have in Grafana later today - afk for most of the day.  Most likely a demonstration of my lack of influxdb expertise.  In any case even if it’s real the tag can easily be put behind a pref.

**Comment by iamtrep** [comment]
Took a bit of a break but finally got back to this.   

So what I am seeing is that all tags taken together define a sequence within a measurement.  If nothing is done, what I observe in Grafana is that the measurement will change colour at the time a new tag is added to it.   If I look at the table output, there's indeed two tables there, one with the new "type" tag, the other without.   There's an easy way to deal with this in Flux, using the drop() function to drop the column corresponding to the new tag: a single table is then returned and the graph looks identical to what it did before (there's also the converse, keep(), which if used drops new tags if any appear).

The point being : a current user of InfluxDB-Logger, upon updating to a version with this PR integrated, depending on their Grafana setup, might see their graphs change colour at time of update.  

Doesn't seem like a huge problem to me (it's transient anyway). 

Hope it's clearer.  Let me know.

**Comment by iamtrep** [comment]
Please let me know if you plan to accept this PR.

**Comment by dennypage** [comment]
@iamtrep, I'd like to talk through this again.

Things I'm thinking about:
* With the soft poll changes, is tagging virtual events still important?
* If InfluxDB fixed the missing data problem, would there still be value to having virtual events?
* Does it make sense to tag real events when soft polling is disabled?
* Related, is the tag necessary on both real and virtual events, or could it be virtual events only?
* Will people actually use the tag in Grafana? If so, how will it be used?

**Comment by iamtrep** [comment]
Thanks for the followup.  Fake events are still wrong even if there are fewer but I can do without this PR by using separate instances.  Let's rediscuss if/when other users request this.
